patch/cisco-pmbus-driver-add-support-for-additional-temp-sensor.patch

```diff
@@ -0,0 +1,116 @@
+From aa09f80eb343110869b796205a45c16ed8951dea Mon Sep 17 00:00:00 2001
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Tue, 1 Jun 2021 19:31:36 -0700
+Subject: [PATCH] hwmon: (pmbus) Support fourth PSU temperature sensor
+
+PSU650W has 4 temperature sensors and pmbus driver has support
+for 3 sensors. Added support for 4th temp sensor.
+(i.e. PSU Outlet temperature sensor)
+
+PSU650W is based on LITE-ON vendor.
+LITE-ON MFG part numbers for the PSU are
+PS-2651-3SB5 Z and PS-2651-3SA5 Z.
+
+Signed-off-by: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+---
+ drivers/hwmon/pmbus/pmbus.c      |  4 +++-
+ drivers/hwmon/pmbus/pmbus.h      |  3 +++
+ drivers/hwmon/pmbus/pmbus_core.c | 38 ++++++++++++++++++++++++++++++++
+ 3 files changed, 44 insertions(+), 1 deletion(-)
+
+diff --git a/drivers/hwmon/pmbus/pmbus.c b/drivers/hwmon/pmbus/pmbus.c
+index 4db0400b7..1ce0e00d0 100644
+--- a/drivers/hwmon/pmbus/pmbus.c
++++ b/drivers/hwmon/pmbus/pmbus.c
+@@ -66,8 +66,10 @@ static void pmbus_find_sensor_groups(struct i2c_client *client,
+ 		info->func[0] |= PMBUS_HAVE_TEMP2;
+ 	if (pmbus_check_word_register(client, 0, PMBUS_READ_TEMPERATURE_3))
+ 		info->func[0] |= PMBUS_HAVE_TEMP3;
++	if (pmbus_check_word_register(client, 0, PMBUS_READ_TEMPERATURE_4))
++		info->func[0] |= PMBUS_HAVE_TEMP4;
+ 	if (info->func[0] & (PMBUS_HAVE_TEMP | PMBUS_HAVE_TEMP2
+-			     | PMBUS_HAVE_TEMP3)
++			     | PMBUS_HAVE_TEMP3 | PMBUS_HAVE_TEMP4)
+ 	    && pmbus_check_byte_register(client, 0,
+ 					 PMBUS_STATUS_TEMPERATURE))
+ 			info->func[0] |= PMBUS_HAVE_STATUS_TEMP;
+diff --git a/drivers/hwmon/pmbus/pmbus.h b/drivers/hwmon/pmbus/pmbus.h
+index 5481ff9f8..cdeb3bbee 100644
+--- a/drivers/hwmon/pmbus/pmbus.h
++++ b/drivers/hwmon/pmbus/pmbus.h
+@@ -130,6 +130,8 @@ enum pmbus_regs {
+ 	PMBUS_MFR_DATE			= 0x9D,
+ 	PMBUS_MFR_SERIAL		= 0x9E,
+ 
++	PMBUS_READ_TEMPERATURE_4	= 0xDF,
```

Comment:
Reading through `drivers/hwmon/pmbus/pmbus.h`, is that a standard PMBus register, or should it be a “virtual register”?

     * Virtual registers.
     * Useful to support attributes which are not supported by standard PMBus
     * registers but exist as manufacturer specific registers on individual chips.
     * Must be mapped to real registers in device specific code.

Reply:
Hi Paul,

I was reading up on pmbus commands.
Please check the following link page 103,
https://470q2hhkn9g15l4bc2btbal1-wpengine.netdna-ssl.com/wp-content/uploads/2021/05/PMBus-Specification-Rev-1-2-Part-II-20100906.pdf

According to the specifications the 0xDF is Manufacturer specific command. These are not Virtual registers.

Reply:
Thank you for looking this up. As it’s manufacturer specific, you cannot put it into the “general” core driver and define it as temperature sensor 4. Run `git grep -i 0xdf drivers/hwmon/pmbus/` to see that. I guess, you need to create a dedicated driver for that PSU. There are several dedicated driver in `drivers/hwmon/pmbus/` already. But I am not a PMBus  developer, so am not able to help here further.